File: apps/src/lib/api/transport-web-commands/quota.ts
import type { WebCommandDescriptor } from "./shared";

export function createQuotaWebCommands(): Record<string, WebCommandDescriptor> {
  return {
    service_quota_overview: { rpcMethod: "quota/overview" },
    service_quota_model_usage: { rpcMethod: "quota/modelUsage" },
    service_quota_api_key_usage: { rpcMethod: "quota/apiKeyUsage" },
    service_quota_source_list: { rpcMethod: "quota/sourceList" },
    service_quota_model_pools: { rpcMethod: "quota/modelPools" },
    service_quota_system_pool: { rpcMethod: "quota/systemPool" },
    service_quota_capacity_config: { rpcMethod: "quota/capacityConfig" },
    service_quota_billing_rules: { rpcMethod: "quota/billingRules" },
    service_quota_billing_rule_upsert: {
      rpcMethod: "quota/billingRule/upsert",
      mapParams: (params) => ({
        id: typeof params?.id === "string" ? params.id : null,
        name: typeof params?.name === "string" ? params.name : "",
        status: typeof params?.status === "string" ? params.status : null,
        priority: typeof params?.priority === "number" ? params.priority : null,
        multiplierMillis: typeof params?.multiplierMillis === "number" ? params.multiplierMillis : typeof params?.multiplier_millis === "number" ? params.multiplier_millis : 1000,
        modelPattern: typeof params?.modelPattern === "string" ? params.modelPattern : typeof params?.model_pattern === "string" ? params.model_pattern : null,
        serviceTier: typeof params?.serviceTier === "string" ? params.serviceTier : typeof params?.service_tier === "string" ? params.service_tier : null,
        userId: typeof params?.userId === "string" ? params.userId : typeof params?.user_id === "string" ? params.user_id : null,
        apiKeyId: typeof params?.apiKeyId === "string" ? params.apiKeyId : typeof params?.api_key_id === "string" ? params.api_key_id : null,
        startsAt: typeof params?.startsAt === "number" ? params.startsAt : typeof params?.starts_at === "number" ? params.starts_at : null,
        endsAt: typeof params?.endsAt === "number" ? params.endsAt : typeof params?.ends_at === "number" ? params.ends_at : null,
      }),
    },
    service_quota_billing_rule_delete: { rpcMethod: "quota/billingRule/delete", mapParams: (params) => ({ id: typeof params?.id === "string" ? params.id : "" }) },
    service_quota_source_models_set: {
      rpcMethod: "quota/sourceModels/set",
      mapParams: (params) => ({
        sourceKind: typeof params?.sourceKind === "string" ? params.sourceKind : typeof params?.source_kind === "string" ? params.source_kind : "",
        sourceId: typeof params?.sourceId === "string" ? params.sourceId : typeof params?.source_id === "string" ? params.source_id : "",
        modelSlugs: Array.isArray(params?.modelSlugs) ? params.modelSlugs : Array.isArray(params?.model_slugs) ? params.model_slugs : [],
      }),
    },
    service_quota_capacity_template_update: { rpcMethod: "quota/capacityTemplate/update" },
    service_quota_account_capacity_override_update: { rpcMethod: "quota/accountCapacityOverride/update" },
    service_quota_refresh_sources: { rpcMethod: "quota/refreshSources", mapParams: (params) => ({ kinds: Array.isArray(params?.kinds) ? params.kinds : [], sourceIds: Array.isArray(params?.sourceIds) ? params.sourceIds : Array.isArray(params?.source_ids) ? params.source_ids : [] }) },
  };
}

File: crates/service/src/rpc_dispatch/quota.rs
use codexmanager_core::rpc::types::{JsonRpcRequest, JsonRpcResponse};

use crate::quota::{
    api_key_usage,
    read::{self, BillingRuleUpsertInput, QuotaRefreshSourcesInput},
};

fn string_array_param(req: &JsonRpcRequest, key: &str) -> Vec<String> {
    req.params
        .as_ref()
        .and_then(|value| value.get(key))
        .and_then(|value| value.as_array())
        .map(|items| {
            items
                .iter()
                .filter_map(|item| item.as_str())
                .map(str::trim)
                .filter(|value| !value.is_empty())
                .map(ToString::to_string)
                .collect()
        })
        .unwrap_or_default()
}

pub(super) fn try_handle(req: &JsonRpcRequest) -> Option<JsonRpcResponse> {
    let result = match req.method.as_str() {
        "quota/overview" => super::value_or_error(read::read_quota_overview()),
        "quota/modelUsage" => {
            let start_ts = super::i64_param(req, "startTs");
            let end_ts = super::i64_param(req, "endTs");
            super::value_or_error(read::read_quota_model_usage(start_ts, end_ts))
        }
        "quota/apiKeyUsage" => super::value_or_error(api_key_usage::read_quota_api_key_usage()),
        "quota/sourceList" => super::value_or_error(read::read_quota_source_list()),
        "quota/modelPools" => super::value_or_error(read::read_quota_model_pools()),
        "quota/systemPool" => {
            let reference_model = super::string_param(req, "referenceModel");
            super::value_or_error(read::read_quota_system_pool(reference_model))
        }
        "quota/capacityConfig" => super::value_or_error(read::read_quota_capacity_config()),
        "quota/billingRules" => super::value_or_error(read::read_billing_rules()),
        "quota/billingRule/upsert" => {
            super::value_or_error(read::upsert_billing_rule(BillingRuleUpsertInput {
                id: super::string_param(req, "id"),
                name: super::str_param(req, "name").unwrap_or("").to_string(),
                status: super::string_param(req, "status"),
                priority: super::i64_param(req, "priority"),
                multiplier_millis: super::i64_param(req, "multiplierMillis").unwrap_or(1_000),
                model_pattern: super::string_param(req, "modelPattern"),
                service_tier: super::string_param(req, "serviceTier"),
                user_id: super::string_param(req, "userId"),
                project_id: super::string_param(req, "projectId"),
                api_key_id: super::string_param(req, "apiKeyId"),
                starts_at: super::i64_param(req, "startsAt"),
                ends_at: super::i64_param(req, "endsAt"),
            }))
        }
        "quota/billingRule/delete" => {
            let id = super::str_param(req, "id").unwrap_or("");
            super::value_or_error(read::delete_billing_rule(id))
        }
        "quota/capacityTemplate/update" => {
            let plan_type = super::str_param(req, "planType").unwrap_or("");
            super::value_or_error(read::update_account_quota_capacity_template(
                plan_type,
                super::i64_param(req, "primaryWindowTokens"),
                super::i64_param(req, "secondaryWindowTokens"),
            ))
        }
        "quota/accountCapacityOverride/update" => {
            let account_id = super::str_param(req, "accountId").unwrap_or("");
            super::value_or_error(read::update_account_quota_capacity_override(
                account_id,
                super::i64_param(req, "primaryWindowTokens"),
                super::i64_param(req, "secondaryWindowTokens"),
            ))
        }
        "quota/refreshSources" => {
            super::value_or_error(read::refresh_quota_sources(QuotaRefreshSourcesInput {
                kinds: string_array_param(req, "kinds"),
                source_ids: string_array_param(req, "sourceIds"),
            }))
        }
        _ => return None,
    };

    Some(super::response(req, result))
}
